File: leetcode/残酷刷题/1751. Maximum Number of Events That Can Be Attended II.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
   public:
    int maxValue(vector<vector<int>> &events, int k) {
        int n = events.size();
        sort(events.begin(), events.end(), [](auto &a, auto &b) {
            return a[1] < b[1];
        });
        vector<int> s;
        // 添加哨兵
        events.insert(events.begin(), {0, 0, 0});
        for (auto &e : events) s.push_back(e[1]);
        // f[0][1] ~f[0][k] 均为 0，表示前面可以选人一多，这样就使得最终真正在 1~n 里面选的总数可以小于 k 了
        vector<vector<int>> f(n + 1, vector<int>(k + 1));
        for (int i = 1; i <= n; i++) {
            int j = lower_bound(s.begin(), s.end(), events[i][0]) - 1 - s.begin();
            for (int t = 1; t <= k; t++) {
                f[i][t] = max(f[i - 1][t], f[j][t - 1] + events[i][2]);
            }
        }
        return f[n][k];
    }
};

File: leetcode/残酷刷题/1751. Maximum Number of Events That Can Be Attended II.ts
function maxValue (events: number[][], k: number): number {
/*
f[i][j] 表示参加到第 i 个活动为止，参加了 j 个事件的最大收益
对于每一个事件 i 要找到前面最后的一个收益最大的事件来进行转移

g[i][j] 表示参加到第 <=i 个活动为止，参加了 j 个事件的最大收益
*/
  const n = events.length
  const g = [...Array(n)].map(() => Array<number>(k + 1).fill(0))
  events.sort((a, b) => a[1] - b[1])
  const ans = Array(k + 1).fill(0)

  for (let i = 0; i < n; i++) {
    const [s, e, v] = events[i]
    // 找到最后一个在当前时间开始之前的事件
    let [l, r] = [0, i - 1]
    while (l < r) {
      const mid = l + r + 1 >> 1
      if (events[mid][1] < s) l = mid
      else r = mid - 1
    }
    const prev = (events[l][1] < s) ? l : -1
    for (let j = 1; j <= k; j++) {
      const now = (prev === -1 ? 0 : g[prev][j - 1]) + v
      g[i][j] = ans[j] = Math.max(ans[j], now)
    }
  }

  return Math.max(...ans)
};
